bftengine/src/preprocessor/tests/preprocessor_test.cpp

```diff
@@ -177,67 +178,97 @@ void setUpConfiguration() {
   memset(buf, '5', bufLen);
-  reqInfo.handlePrimaryPreProcessed(PreProcessRequestMsgSharedPtr(), buf, bufLen);
-  for (auto i = 1; i <= numOfRequiredReplies - 1; i++) {
+  for (auto i = 1; i <= numOfRequiredReplies; i++) {
     if (i != numOfRequiredReplies - 1) Assert(reqInfo.getPreProcessingConsensusResult() == CONTINUE);
-    reqInfo.handlePreProcessReplyMsg(preProcessNonPrimary(reqInfo, i, repInfo));
+    reqInfo.handlePreProcessReplyMsg(preProcessNonPrimary(i, repInfo));
   }
+  reqInfo.handlePrimaryPreProcessed(buf, bufLen);
   Assert(reqInfo.getPreProcessingConsensusResult() == COMPLETE);
 }
 
-TEST(requestPreprocessingInfo_test, primaryReplicaPreProcessingRetry) {
-  RequestProcessingInfo reqInfo(numOfReplicas, numOfRequiredReplies, reqSeqNum, ClientPreProcessReqMsgUniquePtr());
+TEST(requestPreprocessingInfo_test, primaryReplicaPreProcessingRetrySucceeds) {
+  RequestProcessingInfo reqInfo(numOfReplicas,
+                                numOfRequiredReplies,
+                                reqSeqNum,
+                                ClientPreProcessReqMsgUniquePtr(),
+                                PreProcessRequestMsgSharedPtr());
   ReplicasInfo repInfo(replicaConfig, true, true);
   memset(buf, '5', bufLen);
-  reqInfo.handlePrimaryPreProcessed(PreProcessRequestMsgSharedPtr(), buf, bufLen);
+  reqInfo.handlePrimaryPreProcessed(buf, bufLen);
   for (auto i = 1; i <= numOfRequiredReplies; i++) {
     if (i != numOfReplicas - 1) Assert(reqInfo.getPreProcessingConsensusResult() == CONTINUE);
     memset(buf, '4', bufLen);
-    reqInfo.handlePreProcessReplyMsg(preProcessNonPrimary(reqInfo, i, repInfo));
+    reqInfo.handlePreProcessReplyMsg(preProcessNonPrimary(i, repInfo));
   }
   Assert(reqInfo.getPreProcessingConsensusResult() == RETRY_PRIMARY);
   memset(buf, '4', bufLen);
-  reqInfo.handlePrimaryPreProcessed(PreProcessRequestMsgSharedPtr(), buf, bufLen);
+  reqInfo.handlePrimaryPreProcessed(buf, bufLen);
+  Assert(reqInfo.getPreProcessingConsensusResult() == COMPLETE);
+}
+
+TEST(requestPreprocessingInfo_test, primaryReplicaDidNotCompletePreProcessingWhileNonPrimariesDid) {
+  RequestProcessingInfo reqInfo(numOfReplicas,
+                                numOfRequiredReplies,
+                                reqSeqNum,
+                                ClientPreProcessReqMsgUniquePtr(),
+                                PreProcessRequestMsgSharedPtr());
+  ReplicasInfo repInfo(replicaConfig, true, true);
+  memset(buf, '5', bufLen);
+  for (auto i = 1; i <= numOfRequiredReplies; i++) {
+    if (i != numOfReplicas - 1) Assert(reqInfo.getPreProcessingConsensusResult() == CONTINUE);
+    memset(buf, '4', bufLen);
+    reqInfo.handlePreProcessReplyMsg(preProcessNonPrimary(i, repInfo));
+  }
+  Assert(reqInfo.getPreProcessingConsensusResult() == CONTINUE);
+  memset(buf, '4', bufLen);
+  reqInfo.handlePrimaryPreProcessed(buf, bufLen);
   Assert(reqInfo.getPreProcessingConsensusResult() == COMPLETE);
```

Comment:
given the test title "primaryReplicaDidNotCompletePreProcessingWhileNonPrimariesDid", does it make sense to have:
`reqInfo.getPreProcessingConsensusResult() == COMPLETE`

Reply:
That's all the idea. Until primary did not complete pre-execution we cannot complete request handling even with f+1 replies collected. We need to verify that the primary pre-execution result is the same that passed the pre-execution consensus, otherwise, we need to retry primary pre-execution.

Reply:
It's probably just a naming problem from my perspective. Actually, what's the difference between COMPLETE and CONTINUE?

Reply:
Before the fix, I did not verify that the primary complete the pre-execution and this was treated as the primary generated a different pre-execution results hash. The function returned with RETRY_PRIMARY status and then crashed as PreProcessRequestMsg was not been saved yet in RequestProcessingInfo and was NULL. Complicated.

Reply:
I see. Which unit test checks this doesn't happen now?

Reply:
primaryReplicaDidNotCompletePreProcessingWhileNonPrimariesDid